File: src/com/aopphp/go/PointcutQueryParserDefinition.java
package com.aopphp.go;

import com.aopphp.go.parser.PointcutLexer;
import com.intellij.lang.ASTNode;
import com.intellij.lang.Language;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lexer.FlexAdapter;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import com.aopphp.go.parser.PointcutParser;
import com.aopphp.go.psi.PointcutFile;
import com.aopphp.go.psi.PointcutTypes;
import org.jetbrains.annotations.NotNull;

import java.io.Reader;

public class PointcutQueryParserDefinition implements ParserDefinition{
    public static final TokenSet WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE);
    public static final TokenSet COMMENTS = TokenSet.create(PointcutTypes.COMMENT);

    public static final IFileElementType FILE = new IFileElementType(Language.<PointcutQueryLanguage>findInstance(PointcutQueryLanguage.class));

    @NotNull
    @Override
    public Lexer createLexer(Project project) {
        return new FlexAdapter(new PointcutLexer((Reader) null));
    }

    @NotNull
    public TokenSet getWhitespaceTokens() {
        return WHITE_SPACES;
    }

    @NotNull
    public TokenSet getCommentTokens() {
        return COMMENTS;
    }

    @NotNull
    public TokenSet getStringLiteralElements() {
        return TokenSet.EMPTY;
    }

    @NotNull
    public PsiParser createParser(final Project project) {
        return new PointcutParser();
    }

    @Override
    public IFileElementType getFileNodeType() {
        return FILE;
    }

    public PsiFile createFile(FileViewProvider viewProvider) {
        return new PointcutFile(viewProvider);
    }

    public SpaceRequirements spaceExistanceTypeBetweenTokens(ASTNode left, ASTNode right) {
        return SpaceRequirements.MAY;
    }

    @NotNull
    public PsiElement createElement(ASTNode node) {
        return PointcutTypes.Factory.createElement(node);
    }
}

File: src/main/kotlin/com/aopphp/go/PointcutQueryParserDefinition.kt
package com.aopphp.go

import com.aopphp.go.parser.PointcutLexer
import com.aopphp.go.parser.PointcutParser
import com.aopphp.go.psi.PointcutFile
import com.aopphp.go.psi.PointcutTypes
import com.intellij.lang.ASTNode
import com.intellij.lang.Language
import com.intellij.lang.ParserDefinition
import com.intellij.lexer.FlexAdapter
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IFileElementType
import com.intellij.psi.tree.TokenSet

class PointcutQueryParserDefinition : ParserDefinition {

    companion object {
        @JvmField val WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE)
        @JvmField val COMMENTS = TokenSet.create(PointcutTypes.T_COMMENT)
        @JvmField val FILE = IFileElementType(Language.findInstance(PointcutQueryLanguage::class.java))
    }

    override fun createLexer(project: Project) = FlexAdapter(PointcutLexer(null))

    override fun getWhitespaceTokens() = WHITE_SPACES

    override fun getCommentTokens() = COMMENTS

    override fun getStringLiteralElements(): TokenSet = TokenSet.EMPTY

    override fun createParser(project: Project) = PointcutParser()

    override fun getFileNodeType() = FILE

    override fun createFile(viewProvider: FileViewProvider) = PointcutFile(viewProvider)

    override fun spaceExistenceTypeBetweenTokens(left: ASTNode, right: ASTNode) =
        ParserDefinition.SpaceRequirements.MAY

    override fun createElement(node: ASTNode) = PointcutTypes.Factory.createElement(node)!!
}
